Possible fix for hotmail related accounts.

mustyoshi opened the pull request:
I was writing a bot, and I noticed it kept crashing in large groups. After a lot of debugging, I discovered that the root cause *seems* to be hotmail accounts which start with "1:". I made the included change, and it seems to work just fine now.

EionRobb commented:
Do you have a backtrace for the crash? It'd be good to fix that as there might be issues with an email address turning up. :) I'll also check what the official client does with them

mustyoshi commented:
This are the errors it would spit out before crashing. Ending in a segfault. I have 5 huge 1-2MB files with the full debug stuff from the sessions if you want to take a look at them.
Here's what one of them looks like in the array from the Skype servers
{
			"id": "1:removed@hotmail.com",
			"type": "ThreadMember",
			"userLink": "https://bn1-client-s.gateway.messenger.live.com/v1/users/1:removed@hotmail.com",
			"role": "User",
			"capabilities": [],
			"linkedMri": "8:live:removed",
			"cid": -9156479350410200091,
			"userTile": "",
			"friendlyName": ""
		}

** (process:25306): CRITICAL **: purple_normalize: assertion 'str != NULL' failed

** (process:25306): CRITICAL **: purple_find_buddy: assertion '(name != NULL) && (*name != '\0')' failed

** (process:25306): CRITICAL **: purple_conv_chat_is_user_ignored: assertion 'user != NULL' failed

** (process:25306): CRITICAL **: purple_conv_chat_cb_new: assertion 'name != NULL' failed

** (process:25306): CRITICAL **: purple_find_buddy: assertion '(name != NULL) && (*name != '\0')' failed

mustyoshi commented:
Valgrind output.

==26211== Invalid write of size 4
==26211==    at 0x4E93EAE: purple_conv_chat_add_users (in /usr/lib/libpurple.so.0.10.9)
==26211==    by 0x4E941DF: purple_conv_chat_add_user (in /usr/lib/libpurple.so.0.10.9)
==26211==    by 0xE984E42: skypeweb_got_thread_users (skypeweb_messages.c:531)
==26211==    by 0xE9820E1: skypeweb_connection_process_data (skypeweb_connection.c:244)
==26211==    by 0xE9820E1: skypeweb_post_or_get_readdata_cb (skypeweb_connection.c:346)
==26211==    by 0x4044DD: purple_glib_io_invoke(_GIOChannel*, GIOCondition, void*) (imclient.c:171)
==26211==    by 0x51BAB6C: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x51BAF47: ??? (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x51BB271: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x405BEC: main (imclient.c:668)
==26211==  Address 0x18 is not stack'd, malloc'd or (recently) free'd
==26211==
==26211==
==26211== Process terminating with default action of signal 11 (SIGSEGV)
==26211==  Access not within mapped region at address 0x18
==26211==    at 0x4E93EAE: purple_conv_chat_add_users (in /usr/lib/libpurple.so.0.10.9)
==26211==    by 0x4E941DF: purple_conv_chat_add_user (in /usr/lib/libpurple.so.0.10.9)
==26211==    by 0xE984E42: skypeweb_got_thread_users (skypeweb_messages.c:531)
==26211==    by 0xE9820E1: skypeweb_connection_process_data (skypeweb_connection.c:244)
==26211==    by 0xE9820E1: skypeweb_post_or_get_readdata_cb (skypeweb_connection.c:346)
==26211==    by 0x4044DD: purple_glib_io_invoke(_GIOChannel*, GIOCondition, void*) (imclient.c:171)
==26211==    by 0x51BAB6C: g_main_context_dispatch (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x51BAF47: ??? (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x51BB271: g_main_loop_run (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
==26211==    by 0x405BEC: main (imclient.c:668)
==26211==  If you believe this happened as a result of a stack
==26211==  overflow in your program's main thread (unlikely but
==26211==  possible), you can try to increase the size of the
==26211==  main thread stack using the --main-stacksize= flag.
==26211==  The main thread stack size used in this run was 10485760.
==26211==
==26211== HEAP SUMMARY:
==26211==     in use at exit: 7,487,222 bytes in 81,246 blocks
==26211==   total heap usage: 657,223 allocs, 575,977 frees, 48,363,905 bytes allocated
==26211==
==26211== LEAK SUMMARY:
==26211==    definitely lost: 474 bytes in 6 blocks
==26211==    indirectly lost: 0 bytes in 0 blocks
==26211==      possibly lost: 2,460,209 bytes in 3,663 blocks
==26211==    still reachable: 5,024,571 bytes in 77,563 blocks
==26211==         suppressed: 0 bytes in 0 blocks
==26211== Rerun with --leak-check=full to see details of leaked memory
==26211==
==26211== For counts of detected and suppressed errors, rerun with: -v
==26211== ERROR SUMMARY: 1891 errors from 19 contexts (suppressed: 0 from 0)
Segmentation fault

EionRobb commented:
Oh, the "linkedMri": "8:live:removed", bit looks interesting.  We should be
picking up on that methinks, and using that instead

On 19 June 2015 at 12:59, Bradley Goodyear <notifications@github.com> wrote:

> Valgrind output.
>
> ==26211== Invalid write of size 4
> ==26211== at 0x4E93EAE: purple_conv_chat_add_users (in
> /usr/lib/libpurple.so.0.10.9)
> ==26211== by 0x4E941DF: purple_conv_chat_add_user (in
> /usr/lib/libpurple.so.0.10.9)
> ==26211== by 0xE984E42: skypeweb_got_thread_users (skypeweb_messages.c:531)
> ==26211== by 0xE9820E1: skypeweb_connection_process_data
> (skypeweb_connection.c:244)
> ==26211== by 0xE9820E1: skypeweb_post_or_get_readdata_cb
> (skypeweb_connection.c:346)
> ==26211== by 0x4044DD: purple_glib_io_invoke(_GIOChannel*, GIOCondition,
> void*) (imclient.c:171)
> ==26211== by 0x51BAB6C: g_main_context_dispatch (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x51BAF47: ??? (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x51BB271: g_main_loop_run (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x405BEC: main (imclient.c:668)
> ==26211== Address 0x18 is not stack'd, malloc'd or (recently) free'd
> ==26211==
> ==26211==
> ==26211== Process terminating with default action of signal 11 (SIGSEGV)
> ==26211== Access not within mapped region at address 0x18
> ==26211== at 0x4E93EAE: purple_conv_chat_add_users (in
> /usr/lib/libpurple.so.0.10.9)
> ==26211== by 0x4E941DF: purple_conv_chat_add_user (in
> /usr/lib/libpurple.so.0.10.9)
> ==26211== by 0xE984E42: skypeweb_got_thread_users (skypeweb_messages.c:531)
> ==26211== by 0xE9820E1: skypeweb_connection_process_data
> (skypeweb_connection.c:244)
> ==26211== by 0xE9820E1: skypeweb_post_or_get_readdata_cb
> (skypeweb_connection.c:346)
> ==26211== by 0x4044DD: purple_glib_io_invoke(_GIOChannel*, GIOCondition,
> void*) (imclient.c:171)
> ==26211== by 0x51BAB6C: g_main_context_dispatch (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x51BAF47: ??? (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x51BB271: g_main_loop_run (in
> /lib/x86_64-linux-gnu/libglib-2.0.so.0.4200.1)
> ==26211== by 0x405BEC: main (imclient.c:668)
> ==26211== If you believe this happened as a result of a stack
> ==26211== overflow in your program's main thread (unlikely but
> ==26211== possible), you can try to increase the size of the
> ==26211== main thread stack using the --main-stacksize= flag.
> ==26211== The main thread stack size used in this run was 10485760.
> ==26211==
> ==26211== HEAP SUMMARY:
> ==26211== in use at exit: 7,487,222 bytes in 81,246 blocks
> ==26211== total heap usage: 657,223 allocs, 575,977 frees, 48,363,905
> bytes allocated
> ==26211==
> ==26211== LEAK SUMMARY:
> ==26211== definitely lost: 474 bytes in 6 blocks
> ==26211== indirectly lost: 0 bytes in 0 blocks
> ==26211== possibly lost: 2,460,209 bytes in 3,663 blocks
> ==26211== still reachable: 5,024,571 bytes in 77,563 blocks
> ==26211== suppressed: 0 bytes in 0 blocks
> ==26211== Rerun with --leak-check=full to see details of leaked memory
> ==26211==
> ==26211== For counts of detected and suppressed errors, rerun with: -v
> ==26211== ERROR SUMMARY: 1891 errors from 19 contexts (suppressed: 0 from
> 0)
> Segmentation fault
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EionRobb/skype4pidgin/pull/283#issuecomment-113332454>
> .
>

EionRobb commented:
Want to try with the latest commit and see if that does the trick?

mustyoshi commented:
Just looking at it, if "linkedMri" = "" is enough to return true, it's not going to work, every user has the "linkedMri" = "", except the Live users have it set to something other than a null string. But I'm compiling it atm.

mustyoshi commented:
It seems to work. Guess I was wrong.

EionRobb commented:
The important bit to know though, is do the Live users show up in the chat room user list and can you see their messages?

mustyoshi commented:
That might be harder for me to test...
I'm not exactly using this plugin in pidgin... Let me get back to you on this.

EionRobb commented:
:D alrighty

mustyoshi commented:
So I've only got one live account to message my bot, and it worked just fine. But they are a 8: account, using Skype and looking at the group, I can see only a few 1: accounts. I can not message them, nor add them as contacts.
So I think this might actually be a glitch that they are even in the group. Searching for the hotmail account leads to the account with the name from the linkedMri field.
So yeah... This might just be an edge case, but with the possibility to crash the plugin if encountered.

EionRobb commented:
Sorry I never merged this one in :(

I did add in support for MSN buddies instead in a recent-ish commit instead
